### PR TITLE
fix: Handle many items in category nav menu

### PR DESCRIPTION
**Pull request by Zeyber**
Closes: #1352

### CODE REVIEW

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 1]
```diff
@@ -1,12 +1,56 @@
...
+      // Column header node for each subcategory
+      const columnHeader: NavigationNode = {
+        isHeader: true,
+        title: children[i].title,
+        url: children[i].url,
+      };
+
+      // Split subcategory items and header into columns
+      if (children[i].children) {
+        const clonedChildren = children[i].children.slice(0);
```
Comment (marlass) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]: I would change that to spread operator.
```suggestion
        const clonedChildren = [columnHeader, ...children[i].children];
```

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 2]
```diff
@@ -1,12 +1,56 @@
...
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
-export class NavigationUIComponent {
+export class NavigationUIComponent implements OnInit {
   @Input() dropdownMode = 'list';
   @Input() node: NavigationNode;
+  columns;
+
+  getColumns(children) {
+    const columns = new Array();
```
Comment (marlass) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]: Just use `[]`. I didn't see that anywhere else in our codebase.
```suggestion
    const columns = [];
```

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts (hunk 1 above)
Comment (marlass) [reply projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]: You could even add columnHeader in the spread operator.

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 3]
```diff
@@ -1,12 +1,56 @@
...
+      const columnHeader: NavigationNode = {
+        isHeader: true,
+        title: children[i].title,
+        url: children[i].url,
+      };
+
+      // Split subcategory items and header into columns
+      if (children[i].children) {
+        const clonedChildren = children[i].children.slice(0);
+        clonedChildren.unshift(columnHeader);
```
Comment (marlass) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]:
```suggestion
```

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 4]
```diff
@@ -1,12 +1,56 @@
...
+        while (clonedChildren.length > 0) {
+          columns.push(clonedChildren.splice(0, COLUMN_SIZE));
+        }
+      } else {
+        // Push header only in case of no subcategory items
+        const headerArray = new Array();
+        headerArray.push(columnHeader);
+        columns.push(headerArray);
```
Comment (marlass) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]: I would just do this in one line and remove extra variable.
```suggestion
        columns.push([columnHeader]);
```

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 5]
```diff
@@ -1,12 +1,56 @@
...
+      if (children[i].children) {
+        const clonedChildren = children[i].children.slice(0);
+        clonedChildren.unshift(columnHeader);
+        while (clonedChildren.length > 0) {
+          columns.push(clonedChildren.splice(0, COLUMN_SIZE));
+        }
+      } else {
+        // Push header only in case of no subcategory items
+        const headerArray = new Array();
+        headerArray.push(columnHeader);
```
Comment (marlass) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]:
```suggestion
```

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 6]
```diff
@@ -1,12 +1,56 @@
...
+      // Split subcategory items and header into columns
+      if (children[i].children) {
+        const clonedChildren = children[i].children.slice(0);
+        clonedChildren.unshift(columnHeader);
+        while (clonedChildren.length > 0) {
+          columns.push(clonedChildren.splice(0, COLUMN_SIZE));
+        }
+      } else {
+        // Push header only in case of no subcategory items
+        const headerArray = new Array();
```
Comment (marlass) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]:
```suggestion
```

**Comment by Pucek9** [comment]
QA done 
image

**Pull request merged**